### PR TITLE
DolphinWX: Add a filename column

### PR DESCRIPTION
**Pull request by ShadowsFriend**
Adds a filename column to DolphinWX. This was requested in issue https://bugs.dolphin-emu.org/issues/8972.

### CODE REVIEW

**Comment by sepalani** [comment]
Please could you provide a screenshot?

File: Source/Core/DolphinWX/GameListCtrl.cpp [hunk 1]
```diff
@@ -107,6 +107,12 @@ static int CompareGameListItems(const GameListItem* iso1, const GameListItem* is
 			if (iso1->GetCountry() < iso2->GetCountry())
 				return -1 * t;
 			return 0;
+		case CGameListCtrl::COLUMN_FNAME:
+			if (iso1->GetFileName() > iso2->GetFileName())
+				return  1 * t;
+			if (iso1->GetFileName() < iso2->GetFileName())
+				return -1 * t;
+			return 0;
```
Comment (JosJuice) [inline Source/Core/DolphinWX/GameListCtrl.cpp]: You can use `strcasecmp` here, like the code for `COLUMN_MAKER` and `COLUMN_ID`.

**Comment by JosJuice** [comment]
* Please write `fileName` instead of `fname` in the code.
* It would be useful if you added file name sorting as a fallback for `COLUMN_TITLE` sorting, right after the disc number fallback.

**Comment by ShadowsFriend** [comment]
Thanks for the feedback! Here is the requested screenshot:
dolphinwx_filecolumn

The other stuff is work in progress now. I've also made a mistake in the resizing part, which I fixed locally. In the process, however, I found that if you hide the column left to size, its contents will be pressed into the size column looking like this (this screenshot being from current master without my patch and the region column unticked):
dolphinwx_polluted_size

This probably wasn't found by now because the region column is enabled by default. Because the file column is added to left of size and is hidden by default, this issue would be visible by default with this commit. I suspect autosize of being the culprit. Should I ignore it for now and just work on the file column?

**Comment by ShadowsFriend** [comment]
Just pushed the commit that should implement the requested changes. It also fixes the mistake I made in the resizing part that was mentioned above. It caused the filename column to always be shown even when hidden.
Now there's still that squishing bug from the second screenshot. I found that it doesn't only happen to the size column. It also happens to the filename column when the region column is deactivated.

File: Source/Core/DolphinWX/GameListCtrl.cpp [hunk 2]
```diff
@@ -1265,13 +1278,25 @@ void CGameListCtrl::AutomaticColumnWidth()
...
+		if (resizable > 425)
 		{
-			if (SConfig::GetInstance().m_showMakerColumn)
+			if (SConfig::GetInstance().m_showMakerColumn &&
+				SConfig::GetInstance().m_showFileNameColumn)
+			{
+				SetColumnWidth(COLUMN_TITLE, resizable / 3);
+				SetColumnWidth(COLUMN_MAKER, resizable / 3);
+				SetColumnWidth(COLUMN_FILENAME, resizable / 3);
+			}
```
Comment (sepalani) [inline Source/Core/DolphinWX/GameListCtrl.cpp]: A space is missing between "else if" and the parenthesis. Idem for the next "else if" block

Comment (ShadowsFriend) [inline Source/Core/DolphinWX/GameListCtrl.cpp]: Now that's embarrassing...Thank you. Latest commit corrected it.

**Comment by JosJuice** [comment]
@dolphin-emu-bot rebuild

**Comment by ShadowsFriend** [comment]
In the meantime I found a solution for the bug shown on the second screenshot. Should I add it as a commit to this PR or should I open a different one? Thing is, one line depends on whether this PR will be accepted or not.

**Comment by JosJuice** [comment]
Depends on whether you're comfortable with fixing merge conflicts for either PR when the other is merged, and whether not having the fix will cause problems for this PR. As it is now, it does cause problems. However, I think the File column would fit better visually to the right of Maker, so that all of the long text fields are in one place. In that case, it wouldn't be more of a problem than it was without this PR.

**Comment by ShadowsFriend** [comment]
I've chosen this position only because for me file name and size kind of belong together. But I totally see your point in it being more visually pleasing to right of Maker. I'm going to move it. This hides the issue at least partially. It does, however, reappear when the file column is activated and the Maker column gets autohidden because of the window being too small.

**Comment by ShadowsFriend** [comment]
It's done.
dolphin_fname2

**Comment by JosJuice** [comment]
@dolphin-emu-bot rebuild

**Comment by ShadowsFriend** [comment]
Just changed strcasecmp to wxStricmp (and therefore included wx/wxcrt.h) to hopefully make the OSX buildbot happy.

File: Source/Core/DolphinWX/GameListCtrl.cpp [hunk 3]
```diff
@@ -95,10 +96,19 @@ static int CompareGameListItems(const GameListItem* iso1, const GameListItem* is
 					return t * (iso1->GetRevision() > iso2->GetRevision() ? 1 : -1);
 				if (iso1->GetDiscNumber() != iso2->GetDiscNumber())
 					return t * (iso1->GetDiscNumber() > iso2->GetDiscNumber() ? 1 : -1);
+				if (wxFileNameFromPath(iso1->GetFileName()) !=
+				    wxFileNameFromPath(iso2->GetFileName()))
+				{
+					return t * wxStricmp(wxFileNameFromPath(iso1->GetFileName()),
+					                     wxFileNameFromPath(iso2->GetFileName()));
```
Comment (lioncash) [inline Source/Core/DolphinWX/GameListCtrl.cpp]: You can de-duplicate these function calls by storing to variables.

File: Source/Core/DolphinWX/GameListCtrl.cpp [hunk 4]
```diff
@@ -361,52 +373,85 @@ static wxString NiceSizeFormat(u64 _size)
...
-	wxString name = StrToWxStr(rISOFile.GetName());
-
-	int disc_number = rISOFile.GetDiscNumber() + 1;
-	if (disc_number > 1 && name.Lower().find(wxString::Format("disc %i", disc_number)) == std::string::npos
-	                    && name.Lower().find(wxString::Format("disc%i", disc_number)) == std::string::npos)
+	// Iterate over all columns and fill them with content if they are visible
+	for (int i = 1; i < NUMBER_OF_COLUMN; i++)
 	{
-		name = wxString::Format(_("%s (Disc %i)"), name.c_str(), disc_number);
+		if(GetColumnWidth(i) != 0)
```
Comment (lioncash) [inline Source/Core/DolphinWX/GameListCtrl.cpp]: space after the `if`

File: Source/Core/DolphinWX/GameListCtrl.cpp [hunk 5]
```diff
@@ -1265,12 +1313,27 @@ void CGameListCtrl::AutomaticColumnWidth()
...
 		// We hide the Maker column if the window is too small
-		if (resizable > 400)
+		// Use ShowColumn() instead of SetColumnWidth because
+		// the maker column may have been autohidden and the
+		// therefore the content needs to be restored.
+		if (resizable > 425)
 		{
-			if (SConfig::GetInstance().m_showMakerColumn)
+			if (SConfig::GetInstance().m_showMakerColumn &&
+				SConfig::GetInstance().m_showFileNameColumn)
```
Comment (lioncash) [inline Source/Core/DolphinWX/GameListCtrl.cpp]: uneven indentation

**Comment by ShadowsFriend** [comment]
I've decided to add the fix for the squeezing issue as first commit and to adjust the filename column changes to be based upon it.

File: Source/Core/DolphinWX/GameListCtrl.cpp [hunk 6]
```diff
@@ -615,8 +660,11 @@ void CGameListCtrl::ScanForISOs()
 
 void CGameListCtrl::OnColBeginDrag(wxListEvent& event)
 {
-	if (event.GetColumn() != COLUMN_TITLE && event.GetColumn() != COLUMN_MAKER)
+	if (event.GetColumn() != COLUMN_TITLE && event.GetColumn() != COLUMN_MAKER &&
+	    event.GetColumn() != COLUMN_FILENAME)
+	{
 		event.Veto();
+	}
 }
```
Comment (lioncash) [inline Source/Core/DolphinWX/GameListCtrl.cpp]:
```cpp
 void CGameListCtrl::OnColBeginDrag(wxListEvent& event)
 {
	const int column_id = event.GetColumn();

	if (column_id != COLUMN_TITLE && column_id != COLUMN_MAKER && column_id != COLUMN_FILENAME)
 		event.Veto();
 }
```

**Comment by ShadowsFriend** [comment]
Thank you for the feedback! The changes have been implemented.

**Comment by degasus** [comment]
LGTM

**Pull request merged**